typescript/sdk/src/core/HyperlaneRelayer.ts

```diff
@@ -206,44 +276,92 @@ export class HyperlaneRelayer {
-      const destination = message.parsed.destination;
-      const chain = this.multiProvider.tryGetChainName(destination);
-      if (!chain) {
-        this.logger.warn(`Unknown destination ${destination}`);
-        return;
+      if (!backlogMsg) {
+        this.logger.trace('Backlog empty, waiting 1s');
+        await sleep(1000);
+        continue;
       }
 
-      if (!chains.includes(chain)) {
-        this.logger.info(`Skipping message to chain ${chain}`);
-        return;
+      // linear backoff (attempts * retryTimeout)
+      if (
+        Date.now() <
+        backlogMsg.lastAttempt + backlogMsg.attempts * this.retryTimeout
+      ) {
+        this.backlog.push(backlogMsg);
+        continue;
       }
 
-      const dispatchReceipt = await event.getTransactionReceipt();
-      const processReceipt = await this.relayMessage(
-        dispatchReceipt,
-        undefined,
-        message,
-      );
+      const { message, dispatchTx, attempts } = backlogMsg;
+      const id = messageId(message);
+      const parsed = parseMessage(message);
+      const dispatchMsg = { id, message, parsed };
+
+      try {
+        const dispatchReceipt = await this.multiProvider
+          .getProvider(parsed.origin)
+          .getTransactionReceipt(dispatchTx);
+
+        // TODO: handle batching
+        await this.relayMessage(dispatchReceipt, undefined, dispatchMsg);
+      } catch (error) {
+        this.logger.error(
+          `Failed to relay message ${id} (attempt #${attempts + 1})`,
+        );
+        this.backlog.push({
+          ...backlogMsg,
+          attempts: attempts + 1,
+          lastAttempt: Date.now(),
+        });
+      }
+    }
+  }
 
-      this.logger.info(
-        `Message ${message.id} was processed in ${
-          this.multiProvider.tryGetExplorerTxUrl(destination, {
-            hash: processReceipt.transactionHash,
-          }) ?? 'tx ' + processReceipt.transactionHash
-        }`,
-      );
-    }, chains);
+  protected whitelistChains() {
+    return this.whitelist ? Object.keys(this.whitelist) : undefined;
+  }
+
+  start(): void {
+    assert(!this.stopRelayingHandler, 'Relayer already started');
+
+    this.backlog = this.cache?.backlog ?? [];
+
+    const { removeHandler } = this.core.onDispatch(async (message, event) => {
+      if (
+        this.whitelist &&
+        !messageMatchesWhitelist(this.whitelist, message.parsed)
+      ) {
+        this.logger.debug(
+          { message, whitelist: this.whitelist },
+          `Skipping message ${message.id} not matching whitelist`,
+        );
+        return;
+      }
+
+      this.backlog.push({
+        attempts: 0,
+        lastAttempt: Date.now(),
+        message: message.message,
+        dispatchTx: event.transactionHash,
+      });
+    }, this.whitelistChains());
 
     this.stopRelayingHandler = removeHandler;
+
+    // start flushing backlog
+    void this.flushBacklog();
```

Comment:
this might be an antipattern

Reply:
it sure is! Did you do this to swallow the promise error? At least `.catch` it in that case. But more generally speaking, I think the design of this class should be different (see above)

Reply:
this isnt for swallowing the error, its for marking the promise explicitly as hanging (this is what the linter suggests btw)